internal/observability/event/telemetry_filter.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package event
+
+import (
+	"errors"
+	"reflect"
+
+	"google.golang.org/protobuf/proto"
+)
+
+// protoFilter is a signature for a struct field validation test
+type protoFilter func(field reflect.StructField) bool
+
+// telemetryFilter checks a struct field should be included in observation telemetry data
+func telemetryFilter(field reflect.StructField) bool {
+	if field.Tag.Get("eventstream") == "observation" {
+		// log.Printf("%s is allowed by Tags (%s:\"%s\")", name, tag, tagValue)
+		return true
+	}
+	return false
+}
+
+// filterValue will preserve or zero a value based on if it is classed as observable
+func filterValue(fv reflect.Value, isObservable bool) {
+	if isObservable {
+		return // let data persist to telemetry
+	}
+
+	// check for nil value (prevent panics)
+	if fv == reflect.ValueOf(nil) {
+		return
+	}
+
+	if fv.Kind() == reflect.Ptr {
+		fv = fv.Elem()
+	}
+
+	// check to see if it's an exported struct field
+	if !fv.CanSet() {
+		return
+	}
+
+	fv.SetZero()
+
+	return
+}
+
+func recurseStructureWithProtoFilter(value reflect.Value, filterFunc protoFilter, isObservable bool) error {
+	kind := value.Kind()
+
+	switch kind {
+	case reflect.Interface, reflect.Ptr:
+		value = value.Elem()
+		return recurseStructureWithProtoFilter(value, filterFunc, isObservable)
+	case reflect.Map:
+		m := reflect.ValueOf(value)
+		for _, k := range m.MapKeys() {
+			mVal := m.MapIndex(k)
+			if err := recurseStructureWithProtoFilter(mVal, filterFunc, isObservable); err != nil {
+				return err
+			}
+		}
+		return nil
+	case reflect.Array, reflect.Slice:
+		if isObservable {
+			for i := 0; i < value.Len(); i++ {
+				sVal := value.Index(i)
+				if err := recurseStructureWithProtoFilter(sVal, filterFunc, isObservable); err != nil {
+					return err
+				}
+			}
+		} else {
+			if kind == reflect.Slice {
+				value.SetLen(0) // truncate
+			} else {
+				// fixed size, so we zero
+				for i := 0; i < value.Len(); i++ {
+					value.Index(i).SetZero()
+				}
+			}
+		}
+	case reflect.Struct:
+		fields := value.Type()
+		num := fields.NumField()
+		for i := 0; i < num; i++ {
+			field := fields.Field(i)
+			v := value.Field(i)
+			if !field.IsExported() {
+				continue
+			}
+			isObservable := true
+			if filterFunc != nil {
+				isObservable = filterFunc(field)
+			}
+			if err := recurseStructureWithProtoFilter(v, filterFunc, isObservable); err != nil {
+				return err
+			}
+		}
+		return nil
+	default:
+		// any other non structured type, we will output or not via filterValue
+		filterValue(value, isObservable)
+		return nil
+	}
+
+	return nil
+}
+
+func filterProtoMessage(msg proto.Message, filterFunc protoFilter) (proto.Message, error) {
+	if msg == nil {
+		return nil, errors.New("nil message")
+	}
+
+	cloneMsg := proto.Clone(msg)
+	err := recurseStructureWithProtoFilter(reflect.ValueOf(cloneMsg), filterFunc, false)
+	return cloneMsg, err
+}
```

Comment:
I'm late to the design of this feature, but it seems to me that we could do this better with a proper protobuf option and protobuf reflection rather than using a struct tag and Go reflection. All our protobuf generation uses the Go APIv2, which supports rich reflection of protobuf options on generated structs. See https://go.dev/blog/protobuf-apiv2 for an introduction and a simple example of doing almost exactly what we're doing here, but using a protobuf option instead of a struct tag. Perhaps this was considered already and rejected, but personally I think I'd prefer this to the struct tags that we're using now, which is a bit of a hack (it's not portable to other protobuf languages and requires post-processing of our generated files). There is a precedent for using protobuf options in our protobuf files for custom logic such as this. What do you think?

Reply:
Thank you for the comment, Johan. This is an interesting idea and I was not aware of this before. Tomorrow is our telemetry poodle, I would be really happy to invite you over if you are available and we talk about it. I will start the thread on Slack so we can talk more about it. Quick note that we are pretty close to the deadline to deliver this feature and I am not sure about switching to protobuf options at this time but we can definitely talk about it tomorrow